File: backend/src/main/java/com/main/Difficult/Controllers/ProductoController.java
package com.main.Difficult.Controllers;

import com.fasterxml.jackson.annotation.JsonView;
import com.main.Difficult.Domain.producto.Producto;
import com.main.Difficult.Domain.producto.View;
import com.main.Difficult.Services.ProductoService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@CrossOrigin(origins ={"*"})
public class ProductoController {
    @Autowired private ProductoService productoService;

    @JsonView(View.DetalleDeProducto.class)
    @GetMapping("/producto/{idProducto}")
    public Producto findById(@PathVariable String idProducto){
        return productoService.findById(idProducto);
    }

    @JsonView(View.ProductoLista.class)
    @GetMapping("/producto/traer")
    public List<Producto> buscarPor(
            @RequestParam(value = "nombre" , required = false) String nombre,
            @RequestParam(value = "puntaje", required = false,defaultValue = "0") Integer puntaje,
            @RequestParam(value = "paisDeOrigen" ,required = false,defaultValue = "") String paisDeOrigen
    )
    {
        return productoService.buscarPor(nombre,paisDeOrigen,puntaje);
    }
}

File: backend/src/main/kotlin/aplicacion/controladores/ProductoController.kt
package aplicacion.controladores

import aplicacion.dominio.producto.DatosDeBusqueda
import aplicacion.dominio.producto.Producto
import aplicacion.dominio.producto.View
import aplicacion.servicios.ProductoService
import com.fasterxml.jackson.annotation.JsonView
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.*

@RestController
@CrossOrigin(origins = ["*"], methods = [RequestMethod.GET])
class ProductoController {
    @Autowired private lateinit var productoService: ProductoService

    @JsonView(View.ProductoIndividual::class)
    @GetMapping("/producto/{id}")
    fun findById(
        @PathVariable id: Long
    ): Producto {
        return productoService.findById(id)
    }

    @JsonView(View.ProductoLista::class)
    @GetMapping("/producto/traer")
    fun buscarPor(
        @RequestParam(value = "nombre", required = false) nombre: String?,
        @RequestParam(value = "puntaje", required = false) puntaje: Int?,
        @RequestParam(value = "paisDeOrigen", required = false) paisDeOrigen: String?
    ): List<Producto> {
        val datos = DatosDeBusqueda(nombre, puntaje, paisDeOrigen)
        return productoService.buscarPor(datos)
    }
}